r45555 | kakuto | 2017-10-17 23:52:13 +0200 (Tue, 17 Oct 2017) | 1 line

pdftexdir: remove unused lines in new files
------------------------------------------------------------------------
r45553 | kakuto | 2017-10-17 06:44:29 +0200 (Tue, 17 Oct 2017) | 1 line

pdftexdir: add new files to support poppler 0.59.0 or newer ones (typo)
------------------------------------------------------------------------
r45552 | kakuto | 2017-10-17 06:34:41 +0200 (Tue, 17 Oct 2017) | 1 line

pdftexdir: add new files to support poppler 0.59.0 or new ones (typo)
------------------------------------------------------------------------
r45551 | kakuto | 2017-10-17 06:24:27 +0200 (Tue, 17 Oct 2017) | 1 line

pdftexdir: add new files to support poppler 0.59.0 or newer ones

diff -u source/texk/web2c/pdftexdir/pdftoepdf.cc source/texk/web2c/pdftexdir.newpoppler/pdftoepdf.cc
--- source/texk/web2c/pdftexdir/pdftoepdf.cc	2017-05-08 07:50:08.141501111 +0200
+++ source/texk/web2c/pdftexdir.newpoppler/pdftoepdf.cc	2018-01-19 21:56:11.841887257 +0100
@@ -1,5 +1,5 @@
 /*
-Copyright 1996-2016 Han The Thanh, <thanh@pdftex.org>
+Copyright 1996-2017 Han The Thanh, <thanh@pdftex.org>
 
 This file is part of pdfTeX.
 
@@ -17,6 +17,15 @@
 with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
 
+/*
+This is based on the patch texlive-poppler-0.59.patch <2017-09-19> at
+https://git.archlinux.org/svntogit/packages.git/plain/texlive-bin/trunk
+by Arch Linux. A little modifications are made to avoid a crash for
+some kind of pdf images, such as figure_missing.pdf in gnuplot.
+The poppler should be 0.59.0 or newer versions.
+POPPLER_VERSION should be defined.
+*/
+
 /* Do this early in order to avoid a conflict between
    MINGW32 <rpcndr.h> defining 'boolean' as 'unsigned char' and
    <kpathsea/types.h> defining Pascal's boolean as 'int'.
@@ -39,10 +48,7 @@
 #include <goo/gfile.h>
 #define GString GooString
 #else
-#include <aconf.h>
-#include <GString.h>
-#include <gmem.h>
-#include <gfile.h>
+#error POPPLER_VERSION should be defined.
 #endif
 #include <assert.h>
 
@@ -84,31 +90,6 @@
 #define MASK_SUPPRESS_PTEX_PAGENUMBER 0x04
 #define MASK_SUPPRESS_PTEX_INFODICT   0x08
 
-// PdfObject encapsulates the xpdf Object type,
-// and properly frees its resources on destruction.
-// Use obj-> to access members of the Object,
-// and &obj to get a pointer to the object.
-// It is no longer necessary to call Object::free explicitely.
-
-class PdfObject {
-  public:
-    PdfObject() {               // nothing
-    } ~PdfObject() {
-        iObject.free();
-    }
-    Object *operator->() {
-        return &iObject;
-    }
-    Object *operator&() {
-        return &iObject;
-    }
-  private:                     // no copying or assigning
-    PdfObject(const PdfObject &);
-    void operator=(const PdfObject &);
-  public:
-    Object iObject;
-};
-
 // When copying the Resources of the selected page, all objects are copied
 // recusively top-down. Indirect objects however are not fetched during
 // copying, but get a new object number from pdfTeX and then will be
@@ -212,18 +193,6 @@
     delete pdf_doc;
 }
 
-// Replacement for
-//      Object *initDict(Dict *dict1){ initObj(objDict); dict = dict1; return this; }
-
-static void initDictFromDict(PdfObject & obj, Dict * dict)
-{
-    obj->initDict(xref);
-    for (int i = 0, l = dict->getLength(); i < l; i++) {
-        Object obj1;
-        obj->dictAdd(copyString(dict->getKey(i)), dict->getValNF(i, &obj1));
-    }
-}
-
 // --------------------------------------------------------------------
 
 static int addEncoding(GfxFont * gfont)
@@ -320,10 +289,10 @@
 
 static void copyDictEntry(Object * obj, int i)
 {
-    PdfObject obj1;
+    Object obj1;
     copyName(obj->dictGetKey(i));
     pdf_puts(" ");
-    obj->dictGetValNF(i, &obj1);
+    obj1 = obj->dictGetValNF(i);
     copyObject(&obj1);
     pdf_puts("\n");
 }
@@ -376,17 +345,17 @@
 static void copyProcSet(Object * obj)
 {
     int i, l;
-    PdfObject procset;
+    Object procset;
     if (!obj->isArray())
         pdftex_fail("PDF inclusion: invalid ProcSet array type <%s>",
                     obj->getTypeName());
     pdf_puts("/ProcSet [ ");
     for (i = 0, l = obj->arrayGetLength(); i < l; ++i) {
-        obj->arrayGetNF(i, &procset);
-        if (!procset->isName())
+        procset = obj->arrayGetNF(i);
+        if (!procset.isName())
             pdftex_fail("PDF inclusion: invalid ProcSet entry type <%s>",
-                        procset->getTypeName());
-        copyName(procset->getName());
+                        procset.getTypeName());
+        copyName(procset.getName());
         pdf_puts(" ");
     }
     pdf_puts("]\n");
@@ -394,10 +363,29 @@
 
 #define REPLACE_TYPE1C true
 
+static bool embeddableFont(Object * fontdesc)
+{
+    Object fontfile, ffsubtype;
+
+    if (!fontdesc->isDict())
+        return false;
+    fontfile = fontdesc->dictLookup("FontFile");
+    if (fontfile.isStream())
+        return true;
+    if (REPLACE_TYPE1C) {
+        fontfile = fontdesc->dictLookup("FontFile3");
+        if (!fontfile.isStream())
+            return false;
+        ffsubtype = fontfile.streamGetDict()->lookup("Subtype");
+        return ffsubtype.isName() && !strcmp(ffsubtype.getName(), "Type1C");
+    }
+    return false;
+}
+
 static void copyFont(char *tag, Object * fontRef)
 {
-    PdfObject fontdict, subtype, basefont, fontdescRef, fontdesc, charset,
-        fontfile, ffsubtype, stemV;
+    Object fontdict, subtype, basefont, fontdescRef, fontdesc, charset,
+        stemV;
     GfxFont *gfont;
     fd_entry *fd;
     fm_entry *fontmap;
@@ -413,33 +401,39 @@
     }
     // Only handle included Type1 (and Type1C) fonts; anything else will be copied.
     // Type1C fonts are replaced by Type1 fonts, if REPLACE_TYPE1C is true.
-    if (!fixedinclusioncopyfont && fontRef->fetch(xref, &fontdict)->isDict()
-        && fontdict->dictLookup("Subtype", &subtype)->isName()
-        && !strcmp(subtype->getName(), "Type1")
-        && fontdict->dictLookup("BaseFont", &basefont)->isName()
-        && fontdict->dictLookupNF("FontDescriptor", &fontdescRef)->isRef()
-        && fontdescRef->fetch(xref, &fontdesc)->isDict()
-        && (fontdesc->dictLookup("FontFile", &fontfile)->isStream()
-            || (REPLACE_TYPE1C
-                && fontdesc->dictLookup("FontFile3", &fontfile)->isStream()
-                && fontfile->streamGetDict()->lookup("Subtype",
-                                                     &ffsubtype)->isName()
-                && !strcmp(ffsubtype->getName(), "Type1C")))
-        && (fontmap = lookup_fontmap(basefont->getName())) != NULL) {
+    fontdict = fontRef->fetch(xref);
+    fontdesc = Object(objNull);
+    if (fontdict.isDict()) {
+        subtype = fontdict.dictLookup("Subtype");
+        basefont = fontdict.dictLookup("BaseFont");
+        fontdescRef = fontdict.dictLookupNF("FontDescriptor");
+        if (fontdescRef.isRef()) {
+            fontdesc = fontdescRef.fetch(xref);
+        }
+    }
+    if (!fixedinclusioncopyfont && fontdict.isDict()
+        && subtype.isName()
+        && !strcmp(subtype.getName(), "Type1")
+        && basefont.isName()
+        && fontdescRef.isRef()
+        && fontdesc.isDict()
+        && embeddableFont(&fontdesc)
+        && (fontmap = lookup_fontmap(basefont.getName())) != NULL) {
         // round /StemV value, since the PDF input is a float
         // (see Font Descriptors in PDF reference), but we only store an
         // integer, since we don't want to change the struct.
-        fontdesc->dictLookup("StemV", &stemV);
-        fd = epdf_create_fontdescriptor(fontmap, zround(stemV->getNum()));
-        if (fontdesc->dictLookup("CharSet", &charset) &&
-            charset->isString() && is_subsetable(fontmap))
-            epdf_mark_glyphs(fd, charset->getString()->getCString());
+        stemV = fontdesc.dictLookup("StemV");
+        fd = epdf_create_fontdescriptor(fontmap, zround(stemV.getNum()));
+        charset = fontdesc.dictLookup("CharSet");
+        if (!charset.isNull() &&
+            charset.isString() && is_subsetable(fontmap))
+            epdf_mark_glyphs(fd, charset.getString()->getCString());
         else
             embed_whole_font(fd);
-        addFontDesc(fontdescRef->getRef(), fd);
+        addFontDesc(fontdescRef.getRef(), fd);
         copyName(tag);
         gfont = GfxFont::makeFont(xref, tag, fontRef->getRef(),
-                                  fontdict->getDict());
+                                  fontdict.getDict());
         pdf_printf(" %d 0 R ", addFont(fontRef->getRef(), fd,
                                        addEncoding(gfont)));
     } else {
@@ -451,24 +445,24 @@
 
 static void copyFontResources(Object * obj)
 {
-    PdfObject fontRef;
+    Object fontRef;
     int i, l;
     if (!obj->isDict())
         pdftex_fail("PDF inclusion: invalid font resources dict type <%s>",
                     obj->getTypeName());
     pdf_puts("/Font << ");
     for (i = 0, l = obj->dictGetLength(); i < l; ++i) {
-        obj->dictGetValNF(i, &fontRef);
-        if (fontRef->isRef())
+        fontRef = obj->dictGetValNF(i);
+        if (fontRef.isRef())
             copyFont(obj->dictGetKey(i), &fontRef);
-        else if (fontRef->isDict()) {   // some programs generate pdf with embedded font object
+        else if (fontRef.isDict()) {   // some programs generate pdf with embedded font object
             copyName(obj->dictGetKey(i));
             pdf_puts(" ");
             copyObject(&fontRef);
         }
         else
             pdftex_fail("PDF inclusion: invalid font in reference type <%s>",
-                        fontRef->getTypeName());
+                        fontRef.getTypeName());
     }
     pdf_puts(">>\n");
 }
@@ -557,7 +551,7 @@
 
 static void copyObject(Object * obj)
 {
-    PdfObject obj1;
+    Object obj1;
     int i, l, c;
     Ref ref;
     char *p;
@@ -601,8 +595,8 @@
     } else if (obj->isArray()) {
         pdf_puts("[");
         for (i = 0, l = obj->arrayGetLength(); i < l; ++i) {
-            obj->arrayGetNF(i, &obj1);
-            if (!obj1->isName())
+            obj1 = obj->arrayGetNF(i);
+            if (!obj1.isName())
                 pdf_puts(" ");
             copyObject(&obj1);
         }
@@ -612,9 +606,8 @@
         copyDict(obj);
         pdf_puts(">>");
     } else if (obj->isStream()) {
-        initDictFromDict(obj1, obj->streamGetDict());
         pdf_puts("<<\n");
-        copyDict(&obj1);
+        copyDict(obj->getStream()->getDictObject());
         pdf_puts(">>\n");
         pdf_puts("stream\n");
         copyStream(obj->getStream()->getUndecodedStream());
@@ -638,9 +631,8 @@
     InObj *r;
     for (r = inObjList; r != 0; r = r->next) {
         if (!r->written) {
-            Object obj1;
             r->written = 1;
-            xref->fetch(r->ref.num, r->ref.gen, &obj1);
+            Object obj1 = xref->fetch(r->ref.num, r->ref.gen);
             if (r->type == objFont) {
                 assert(!obj1.isStream());
                 pdfbeginobj(r->num, 2);         // \pdfobjcompresslevel = 2 is for this
@@ -656,7 +648,6 @@
                 pdf_puts("\n");
                 pdfendobj();
             }
-            obj1.free();
         }
     }
 }
@@ -685,7 +676,7 @@
 #ifdef POPPLER_VERSION
         r->font->decRefCnt();
 #else
-        delete r->font;
+#error POPPLER_VERSION should be defined.
 #endif
         delete r;
     }
@@ -728,7 +719,7 @@
 #ifdef POPPLER_VERSION
     int pdf_major_version_found, pdf_minor_version_found;
 #else
-    float pdf_version_found, pdf_version_wanted;
+#error POPPLER_VERSION should be defined.
 #endif
     // initialize
     if (!isInit) {
@@ -760,19 +751,7 @@
         }
     }
 #else
-    pdf_version_found = pdf_doc->doc->getPDFVersion();
-    pdf_version_wanted = 1 + (minor_pdf_version_wanted * 0.1);
-    if (pdf_version_found > pdf_version_wanted + 0.01) {
-        char msg[] =
-            "PDF inclusion: found PDF version <%.1f>, but at most version <%.1f> allowed";
-        if (pdf_inclusion_errorlevel > 0) {
-            pdftex_fail(msg, pdf_version_found, pdf_version_wanted);
-        } else if (pdf_inclusion_errorlevel < 0) {
-            ; /* do nothing */
-        } else { /* = 0, give warning */
-            pdftex_warn(msg, pdf_version_found, pdf_version_wanted);
-        }
-    }
+#error POPPLER_VERSION should be defined.
 #endif
     epdf_num_pages = pdf_doc->doc->getCatalog()->getNumPages();
     if (page_name) {
@@ -839,8 +818,8 @@
     Page *page;
     Ref *pageRef;
     Dict *pageDict;
-    PdfObject contents, obj1, obj2, pageObj, dictObj;
-    PdfObject groupDict;
+    Object contents, obj1, obj2, pageObj, dictObj;
+    Object groupDict;
     bool writeSepGroup = false;
     Object info;
     char *key;
@@ -867,8 +846,8 @@
     encodingList = 0;
     page = pdf_doc->doc->getCatalog()->getPage(epdf_selected_page);
     pageRef = pdf_doc->doc->getCatalog()->getPageRef(epdf_selected_page);
-    xref->fetch(pageRef->num, pageRef->gen, &pageObj);
-    pageDict = pageObj->getDict();
+    pageObj = xref->fetch(pageRef->num, pageRef->gen);
+    pageDict = pageObj.getDict();
     rotate = page->getRotate();
     PDFRectangle *pagebox;
     // write the Page header
@@ -886,7 +865,7 @@
         pdf_printf("/%s.PageNumber %i\n", pdfkeyprefix, (int) epdf_selected_page);
     }
     if ((suppress_ptex_info & MASK_SUPPRESS_PTEX_INFODICT) == 0) {
-        pdf_doc->doc->getDocInfoNF(&info);
+        info = pdf_doc->doc->getDocInfoNF();
         if (info.isRef()) {
             // the info dict must be indirect (PDF Ref p. 61)
             pdf_printf("/%s.InfoDict ", pdfkeyprefix);
@@ -942,14 +921,14 @@
     pdf_puts(stripzeros(s));
 
     // Metadata validity check (as a stream it must be indirect)
-    pageDict->lookupNF("Metadata", &dictObj);
-    if (!dictObj->isNull() && !dictObj->isRef())
+    dictObj = pageDict->lookupNF("Metadata");
+    if (!dictObj.isNull() && !dictObj.isRef())
         pdftex_warn("PDF inclusion: /Metadata must be indirect object");
 
     // copy selected items in Page dictionary except Resources & Group
     for (i = 0; pageDictKeys[i] != NULL; i++) {
-        pageDict->lookupNF(pageDictKeys[i], &dictObj);
-        if (!dictObj->isNull()) {
+        dictObj = pageDict->lookupNF(pageDictKeys[i]);
+        if (!dictObj.isNull()) {
             pdf_newline();
             pdf_printf("/%s ", pageDictKeys[i]);
             copyObject(&dictObj); // preserves indirection
@@ -957,8 +936,8 @@
     } 
 
     // handle page group
-    pageDict->lookupNF("Group", &dictObj);
-    if (!dictObj->isNull()) {
+    dictObj = pageDict->lookupNF("Group");
+    if (!dictObj.isNull()) {
         if (pdfpagegroupval == 0) { 
             // another pdf with page group was included earlier on the
             // same page; copy the Group entry as is.  See manual for
@@ -972,11 +951,36 @@
             copyObject(&dictObj);
         } else {
             // write Group dict as a separate object, since the Page dict also refers to it
-            pageDict->lookup("Group", &dictObj);
-            if (!dictObj->isDict())
+            dictObj = pageDict->lookup("Group");
+            if (!dictObj.isDict())
                 pdftex_fail("PDF inclusion: /Group dict missing");
             writeSepGroup = true;
-            initDictFromDict(groupDict, page->getGroup());
+/*
+This part is only a single line
+            groupDict = Object(page->getGroup());
+in the original patch. In this case, however, pdftex crashes at
+"delete pdf_doc->doc" in "delete_document()" for inclusion of some
+kind of pdf images, for example, figure_missing.pdf in gnuplot.
+A change
+            groupDict = Object(page->getGroup()).copy();
+does not improve the situation.
+The changes below seem to work fine. 
+*/
+// begin modification
+            groupDict = pageDict->lookup("Group");
+            Dict dic1 = page->getGroup();
+            Dict dic2 = groupDict.getDict();
+            // replace dic2 in groupDict with dic1
+            l = dic2.getLength();
+            for (i = 0; i < l; i++) {
+                groupDict.dictRemove(dic2.getKey(i));
+            }
+            l = dic1.getLength();
+            for (i = 0; i < l; i++) {
+                groupDict.dictAdd(copyString(dic1.getKey(i)),
+                                  dic1.getValNF(i));
+            }
+// end modification
             pdf_printf("/Group %ld 0 R\n", (long)pdfpagegroupval);
         }
     }
@@ -989,14 +993,14 @@
         pdftex_warn
             ("PDF inclusion: /Resources missing. 'This practice is not recommended' (PDF Ref)");
     } else {
-        initDictFromDict(obj1, page->getResourceDict());
+        Object *obj1 = page->getResourceDictObject();
         if (!obj1->isDict())
             pdftex_fail("PDF inclusion: invalid resources dict type <%s>",
                         obj1->getTypeName());
         pdf_newline();
         pdf_puts("/Resources <<\n");
         for (i = 0, l = obj1->dictGetLength(); i < l; ++i) {
-            obj1->dictGetVal(i, &obj2);
+            obj2 = obj1->dictGetVal(i);
             key = obj1->dictGetKey(i);
             if (strcmp("Font", key) == 0)
                 copyFontResources(&obj2);
@@ -1009,8 +1013,8 @@
     }
 
     // write the page contents
-    page->getContents(&contents);
-    if (contents->isStream()) {
+    contents = page->getContents();
+    if (contents.isStream()) {
 
         // Variant A: get stream and recompress under control
         // of \pdfcompresslevel
@@ -1021,36 +1025,35 @@
 
         // Variant B: copy stream without recompressing
         //
-        contents->streamGetDict()->lookup("F", &obj1);
-        if (!obj1->isNull()) {
+        obj1 = contents.streamGetDict()->lookup("F");
+        if (!obj1.isNull()) {
             pdftex_fail("PDF inclusion: Unsupported external stream");
         }
-        contents->streamGetDict()->lookup("Length", &obj1);
-        assert(!obj1->isNull());
+        obj1 = contents.streamGetDict()->lookup("Length");
+        assert(!obj1.isNull());
         pdf_puts("/Length ");
         copyObject(&obj1);
         pdf_puts("\n");
-        contents->streamGetDict()->lookup("Filter", &obj1);
-        if (!obj1->isNull()) {
+        obj1 = contents.streamGetDict()->lookup("Filter");
+        if (!obj1.isNull()) {
             pdf_puts("/Filter ");
             copyObject(&obj1);
             pdf_puts("\n");
-            contents->streamGetDict()->lookup("DecodeParms", &obj1);
-            if (!obj1->isNull()) {
+            obj1 = contents.streamGetDict()->lookup("DecodeParms");
+            if (!obj1.isNull()) {
                 pdf_puts("/DecodeParms ");
                 copyObject(&obj1);
                 pdf_puts("\n");
             }
         }
         pdf_puts(">>\nstream\n");
-        copyStream(contents->getStream()->getUndecodedStream());
+        copyStream(contents.getStream()->getUndecodedStream());
         pdfendstream();
-    } else if (contents->isArray()) {
+    } else if (contents.isArray()) {
         pdfbeginstream();
-        for (i = 0, l = contents->arrayGetLength(); i < l; ++i) {
-            Object contentsobj;
-            copyStream((contents->arrayGet(i, &contentsobj))->getStream());
-            contentsobj.free();
+        for (i = 0, l = contents.arrayGetLength(); i < l; ++i) {
+            Object contentsobj = contents.arrayGet(i);
+            copyStream(contentsobj.getStream());
             if (i < l - 1)
                 pdf_newline();  // add a newline after each stream except the last
         }
diff -u source/texk/web2c/pdftexdir/pdftosrc.cc source/texk/web2c/pdftexdir.newpoppler/pdftosrc.cc
--- source/texk/web2c/pdftexdir/pdftosrc.cc	2018-01-19 21:56:11.841887257 +0100
+++ source/texk/web2c/pdftexdir.newpoppler/pdftosrc.cc	2018-01-19 21:56:11.841887257 +0100
@@ -16,6 +16,14 @@
 You should have received a copy of the GNU General Public License along
 with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
+
+/*
+This is based on the patch texlive-poppler-0.59.patch <2017-09-19> at
+https://git.archlinux.org/svntogit/packages.git/plain/texlive-bin/trunk
+by Arch Linux. The poppler should be 0.59.0 or newer versions.
+POPPLER_VERSION should be defined.
+*/
+
 #include <w2c/config.h>
 
 #include <stdlib.h>
@@ -32,10 +40,7 @@
 #include <goo/gmem.h>
 #include <goo/gfile.h>
 #else
-#include <aconf.h>
-#include <GString.h>
-#include <gmem.h>
-#include <gfile.h>
+#error POPPLER_VERSION should be defined.
 #endif
 #include <assert.h>
 
@@ -86,22 +91,20 @@
             objgen = atoi(argv[3]);
     }
     xref = doc->getXRef();
-    catalogDict.initNull();
-    xref->getCatalog(&catalogDict);
+    catalogDict = xref->getCatalog();
     if (!catalogDict.isDict("Catalog")) {
         fprintf(stderr, "No Catalog found\n");
         exit(1);
     }
-    srcStream.initNull();
+    srcStream = Object(objNull);
     if (objnum == 0) {
-        catalogDict.dictLookup("SourceObject", &srcStream);
+        srcStream = catalogDict.dictLookup("SourceObject");
         static char const_SourceFile[] = "SourceFile";
         if (!srcStream.isStream(const_SourceFile)) {
             fprintf(stderr, "No SourceObject found\n");
             exit(1);
         }
-        srcName.initNull();
-        srcStream.getStream()->getDict()->lookup("SourceName", &srcName);
+        srcName = srcStream.getStream()->getDict()->lookup("SourceName");
         if (!srcName.isString()) {
             fprintf(stderr, "No SourceName found\n");
             exit(1);
@@ -110,7 +113,7 @@
         // We cannot free srcName, as objname shares its string.
         // srcName.free();
     } else if (objnum > 0) {
-        xref->fetch(objnum, objgen, &srcStream);
+        srcStream = xref->fetch(objnum, objgen);
         if (!srcStream.isStream()) {
             fprintf(stderr, "Not a Stream object\n");
             exit(1);
@@ -159,29 +162,26 @@
                 int localOffset = 0;
                 Guint firstOffset;
 
-                assert(xref->fetch(e->offset, 0, &objStr)->isStream());
-                nObjects = objStr.streamGetDict()->lookup("N", &obj1)->getInt();
-                obj1.free();
-                first = objStr.streamGetDict()->lookup("First", &obj1)->getInt();
-                obj1.free();
+                objStr = xref->fetch(e->offset, 0);
+                assert(objStr.isStream());
+                obj1 = objStr.streamGetDict()->lookup("N");
+                nObjects = obj1.getInt();
+                obj1 = objStr.streamGetDict()->lookup("First");
+                first = obj1.getInt();
                 firstOffset = objStr.getStream()->getBaseStream()->getStart() + first;
 
                 // parse the header: object numbers and offsets
                 objStr.streamReset();
-                obj1.initNull();
-                str = new EmbedStream(objStr.getStream(), &obj1, gTrue, first);
+                str = new EmbedStream(objStr.getStream(), Object(objNull), gTrue, first);
                 parser = new Parser(xref, new Lexer(xref, str), gFalse);
                 for (n = 0; n < nObjects; ++n) {
-                    parser->getObj(&obj1);
-                    parser->getObj(&obj2);
+                    obj1 = parser->getObj();
+                    obj2 = parser->getObj();
                     if (n == e->gen)
                         localOffset = obj2.getInt();
-                    obj1.free();
-                    obj2.free();
                 }
                 while (str->getChar() != EOF) ;
                 delete parser;
-                objStr.free();
 
                 fprintf(outfile, "%.10lu 00000 n\n",
                         (long unsigned)(firstOffset + localOffset));
@@ -192,7 +192,6 @@
         s->reset();
         while ((c = s->getChar()) != EOF)
             fputc(c, outfile);
-        srcStream.free();
     }
     if (objnum == 0)
         fprintf(stderr, "Source file extracted to %s\n", outname);
@@ -201,7 +200,6 @@
     else
         fprintf(stderr, "Cross-reference table extracted to %s\n", outname);
     fclose(outfile);
-    catalogDict.free();
     delete doc;
     delete globalParams;
 }
